pkg/generic/thrift/http.go

```diff
@@ -77,12 +108,83 @@ func (r *ReadHTTPResponse) SetBase64Binary(enable bool) {
 	r.base64Binary = enable
 }
 
+// SetUseRawBodyForHTTPResp ...
+func (r *ReadHTTPResponse) SetUseRawBodyForHTTPResp(useRawBodyForHTTPResp bool) {
+	r.useRawBodyForHTTPResp = useRawBodyForHTTPResp
+}
+
+// SetDynamicGo ...
+func (r *ReadHTTPResponse) SetDynamicGo(cv t2j.BinaryConv, msg remote.Message) {
+	r.cv = cv
+	r.msg = msg
+	r.dynamicgoEnabled = true
+}
+
 // Read ...
 func (r *ReadHTTPResponse) Read(ctx context.Context, method string, in thrift.TProtocol) (interface{}, error) {
-	fnDsc, err := r.svc.LookupFunctionByMethod(method)
+	// fallback logic
+	if !r.dynamicgoEnabled {
+		fnDsc, err := r.svc.LookupFunctionByMethod(method)
+		if err != nil {
+			return nil, err
+		}
+		fDsc := fnDsc.Response
+		resp, err := skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: r.base64Binary})
+		if r.useRawBodyForHTTPResp {
+			if httpResp, ok := resp.(*descriptor.HTTPResponse); ok && httpResp.Body != nil {
+				rawBody, err := customJson.Marshal(httpResp.Body)
+				if err != nil {
+					return nil, err
+				}
+				httpResp.RawBody = rawBody
+			}
+		}
+		return resp, err
+	}
+
+	// dynamicgo logic
+	if r.msg.PayloadLen() == 0 {
+		return nil, perrors.NewProtocolErrorWithMsg("msg.PayloadLen should always be greater than zero")
+	}
+
+	tProt, ok := in.(*cthrift.BinaryProtocol)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+	}
+
+	tyDsc := r.svc.DynamicGoDsc.Functions()[method].Response()
+
+	mBeginLen := bthrift.Binary.MessageBeginLength(method, thrift.TMessageType(r.msg.MessageType()), r.msg.RPCInfo().Invocation().SeqID())
+	sName, err := in.ReadStructBegin()
+	if err != nil {
+		return nil, err
+	}
+	sBeginLen := bthrift.Binary.StructBeginLength(sName)
+	fName, typeId, id, err := in.ReadFieldBegin()
 	if err != nil {
 		return nil, err
 	}
-	fDsc := fnDsc.Response
-	return skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: r.base64Binary})
+	fBeginLen := bthrift.Binary.FieldBeginLength(fName, typeId, id)
+	transBuf, err := tProt.ByteBuffer().ReadBinary(r.msg.PayloadLen() - mBeginLen - sBeginLen - fBeginLen - bthrift.Binary.MessageEndLength())
+	if err != nil {
+		return nil, err
+	}
+	fid := dthrift.FieldID(id)
+
+	buf := make([]byte, 0, len(transBuf)*2)
+	resp := descriptor.NewHTTPResponse()
+	ctx = context.WithValue(ctx, conv.CtxKeyHTTPResponse, resp)
+
+	for _, field := range tyDsc.Struct().Fields() {
+		if fid == field.ID() {
+			// decode with dynamicgo
+			// thrift []byte to json []byte
+			if err = r.cv.DoInto(ctx, field.Type(), transBuf, &buf); err != nil {
+				return nil, err
+			}
+			break
```

Comment:
Is it that every xxxResult will have only one field xxxResponse, so we break here?

Reply:
@felix021 I think there are 2 possible fields, xxxResponse and xxxException.

Reply:
@felix021 we want to read only 1 field, which meets `fid == field.ID()` on L166. `fid` is the field id retrieved on read. `field.ID()` is the field id from the descriptor.

Reply:
ok, and could it happen that the corresponding field is not found?

Reply:
@felix021 I think it almost never happens. Here, the field id should 0(normal resp id) or 1(exception field id). When the corresponding field is not found, it means that the field id read from the written buf is wrong (= something wrong in the writing process), or the descriptor is not created correctly. Do you think we should check and return error here?

Reply:
It's ok if you're sure it won't happen at all.

I prefer defensive programming especially when it has no impact on performance.